labelbox/schema/bulk_import_request.py

```diff
@@ -296,17 +313,437 @@ def create_from_local_file(cls,
+            annotation.validate_instance(data_row_ids, feature_schemas)
+            uuid = str(annotation.uuid)
+            if uuid in uids:
+                raise labelbox.exceptions.UuidError(
+                    f'{uuid} already used in this import job, '
+                    'must be unique for the project.')
+            uids.add(uuid)
+        except (ValidationError, ValueError, TypeError, KeyError) as e:
+            raise labelbox.exceptions.ValidationError(
+                f"Invalid NDJson on line {idx}") from e
+
+
+#The rest of this file contains objects for MAL validation
+def parse_classification(tool):
+    """
+    Parses a classification from an ontology. Only radio, checklist, and text are supported for mal
+
+    Args:
+        tool (dict)
+
+    Returns:
+        dict
+    """
+    if tool['type'] in ['radio', 'checklist']:
+        return {
+            'tool': tool['type'],
+            'featureSchemaId': tool['featureSchemaId'],
+            'options': [r['featureSchemaId'] for r in tool['options']]
+        }
+    elif tool['type'] == 'text':
+        return {
+            'tool': tool['type'],
+            'featureSchemaId': tool['featureSchemaId']
+        }
+
 
-        - verifies that uuids are unique
+def get_mal_schemas(ontology):
+    """
+    Converts a project ontology to a dict for easier lookup during ndjson validation
 
+    Args:
+        ontology (Ontology)
+    Returns:
+        Dict : Useful for looking up a tool from a given feature schema id
     """
-    uuids: Set[str] = set()
-    for line in lines:
-        uuid = line['uuid']
-        if uuid in uuids:
-            raise labelbox.exceptions.UuidError(
-                f'{uuid} already used in this import job, '
-                'must be unique for the project.')
-        uuids.add(uuid)
+
+    valid_feature_schemas = {}
+    for tool in ontology.normalized['tools']:
+        classifications = [
+            parse_classification(classification_tool)
+            for classification_tool in tool['classifications']
+        ]
+        classifications = {v['featureSchemaId']: v for v in classifications}
+        valid_feature_schemas[tool['featureSchemaId']] = {
+            'tool': tool['tool'],
+            'classifications': classifications
+        }
+    for tool in ontology.normalized['classifications']:
+        valid_feature_schemas[tool['featureSchemaId']] = parse_classification(
+            tool)
+    return valid_feature_schemas
+
+
+LabelboxID: str = pydantic.Field(..., min_length=25, max_length=25)
+
+
+class Bbox(TypedDict):
+    top: float
+    left: float
+    height: float
+    width: float
+
+
+class Point(TypedDict):
+    x: float
+    y: float
+
+
+class FrameLocation(TypedDict):
+    end: int
+    start: int
+
+
+class VideoSupported(BaseModel):
+    #Note that frames are only allowed as top level inferences for video
+    frames: Optional[List[FrameLocation]]
+
+
+class UnionConstructor:
```

Comment:
I'll add some comments to make it more clear in the code. Basically unions don't produce friendly error messages because there is no way to know which of the classes the user intended to construct. What this does is check for specific keys to know what the user intended. If none of those special keys exist then there is another error message alerting the user that they are missing the identifying keys.

Reply:
This object also has some weird functions to allow pydantic to make use of it.